File: pat1004.cpp
/*
	思路：1.用结构体node存储输入数据
	      2.在dfs遍历过程记录每层叶节点个数以及树最大高度
*/


#include<iostream>
#include<vector>
using namespace std;

struct node{
	vector<int> children;
}Tree[100];

int max_depth = 0;
int generation_pop[100] = { 0 };
void dfs(int root, int depth) {
	int len = Tree[root].children.size();
	//	没有孩子节点即节点是叶节点的情况
	if (len == 0) {
		generation_pop[depth] += 1;
		max_depth = depth > max_depth ? depth : max_depth;
	}
	else {
		for (int i = 0; i < len; i++) {
			dfs(Tree[root].children[i], depth + 1);
		}
	}
}

int main() {

	int N, M;
	scanf("%d %d", &N, &M);
	for (int i = 0; i < M; i++) {
		int ID, K, child;
		scanf("%d %d", &ID, &K);
		for (int j = 0; j < K; j++) {
			scanf("%d", &child);
			Tree[ID].children.push_back(child);
		}
	}
	dfs(1, 0);
	printf("%d", generation_pop[0]);
	for (int i = 1; i <= max_depth; i++) {
		printf(" %d", generation_pop[i]);
	}
	return 0;
}

File: pat1004.py
"""
    思路：1.用字典存储输入数据，如{"01":["02"],...}表示01节点有一个孩子节点02
          2.在深度遍历记下每代节点的叶节点数，并求出最大高度
          3.依次输出每代节点的叶节点数

"""


N,M = [int(x) for x in input().split()]
tree = {}
generation_pop = [0] * N    #   初始每代人的数量为0
max_depth = 0              #    最大高度初始为零
def dfs(root, depth):
    global max_depth
    if root not in tree.keys():     #   如果节点不在tree字典中表示没有孩子节点，列表对应数据+1
        generation_pop[depth] += 1
        max_depth = max(depth, max_depth)   #   更新最大高度
    else:
        for child in tree[root]:
            dfs(child, depth + 1)

for i in range(M):
    children = input().split()
    tree[children[0]] = children[2:]

dfs("01",0)
print(generation_pop[0], end='')
for num in generation_pop[1:max_depth+1]:
    print(" " + str(num), end='')